src/providers/VoiceFocusProvider/index.tsx

```diff
@@ -0,0 +1,170 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import React, {
+  useState,
+  useEffect,
+  useRef,
+  useMemo,
+  useContext,
+  createContext,
+  useCallback,
+} from 'react';
+import {
+  Device,
+  VoiceFocusTransformDevice,
+  VoiceFocusDeviceTransformer,
+  VoiceFocusSpec,
+  VoiceFocusDeviceOptions,
+} from 'amazon-chime-sdk-js';
+
+import { useMemoCompare } from '../../utils/use-memo-compare';
+
+interface Props {
+  /** Determines how you want Amazon Voice Focus to behave. This spec is used to derive a runtime configuration when a transformer is created. */
+  spec?: VoiceFocusSpec;
+  /**
+  * A set of options that can be supplied when creating an Amazon Voice Focus device.
+  * For more info, you can go to https://aws.github.io/amazon-chime-sdk-js/interfaces/voicefocusdeviceoptions.html
+  */
+  options?: VoiceFocusDeviceOptions;
+}
+
+interface VoiceFocusState {
+  isVoiceFocusSupported: boolean | undefined;
+  addVoiceFocus: (device: Device) => Promise<Device | VoiceFocusTransformDevice>;
+}
+
+const VoiceFocusContext = createContext<VoiceFocusState | null>(null);
+
+const VoiceFocusProvider: React.FC<Props> = ({
+  spec,
+  options,
+  children
+}) => {
+  const [isVoiceFocusSupported, setIsVoiceFocusSupported] = useState<boolean | undefined>(undefined);
+  const [voiceFocusDevice, setVoiceFocusDevice] = useState<VoiceFocusTransformDevice | null>(null);
+  const [voiceFocusTransformer, setVoiceFocusTransformer] = useState<VoiceFocusDeviceTransformer | null>(null);
+
+  // Check whether the value of spec is undefined or empty object
+  const vfSpec = useMemoCompare(spec, (prev: VoiceFocusSpec | undefined, next: VoiceFocusSpec | undefined): boolean => {
+    if (prev === next) {
+      return true;
+    }
+    if (next === undefined || Object.keys(next).length === 0) {
+      return true;
+    }
+    if (prev === undefined || Object.keys(prev).length === 0) {
+      return true;
+    }
+    return false;
+  });
+
+  const addVoiceFocus = useCallback(async (device: Device): Promise<Device | VoiceFocusTransformDevice> => {
+    if (!voiceFocusDevice) {
+      return createVoiceFocusDevice(device);
+    }
+    const vf = await voiceFocusDevice.chooseNewInnerDevice(device);
+    setVoiceFocusDevice(vf);
+    return vf;
+  }, [isVoiceFocusSupported]);
```

Comment:
Your `useCallback` memoization here isn't quite right. The callback depends on:

1. The device.
2. The transformer.
3. Whether Voice Focus is supported.

By making the dependency list here just `[isVoiceFocusSupported]` you're telling React that it only ever needs to store two devices: one for `false` and one for `true`.

See the big yellow box here:

https://reactjs.org/docs/hooks-reference.html#usecallback

I suspect you don't want to put `useCallback` here at all.